### ISSUE TITLE
Option `--multi-thread` does not make any difference for relatively small repos

### ISSUE THREAD

**Issue by davbeek** [davbeek/gitinspectorgui#38]
In the current implementation, having `--multi-thread` active or not does not make any relevant difference for student repos.

**Comment by davbeek** [comment]
I will set the default to `--no-multi-thread` for now, because it currently makes no difference and it simplifies testing and debugging.

**Comment by davbeek** [comment]
There was no difference for student repos, however for repo main, which has the website of 4TC00, setting `--multi-thread` reduces execution time from 36s to 20s.

**Comment by Alberth289346** [comment]
Not a surprise I think, getting info from Git is I/O bound. Speed is limited by disk or network bandwidth, rather than CPU data processing.

**Comment by davbeek** [comment]
Indeed. Only with option `--blame-history` that generates separate blame info tables per file per commit, for each commit that changes the file, we can get huge html files, where execution speed is limited by CPU html manipulations.

**Comment by Alberth289346** [comment]
At the time we used Python to generate output, the standard tactic was to make lists of lines, and only at the end combine them into a file while writing the output like

``` python
for line in lines:
  write(line)
  write("\n")
```
 (ie thus not `write("\n".join(lines))` since that builds a huge intermediate string)

**Comment by Alberth289346** [comment]
Modern Python are likely better at avoiding these things though.

Try running a profiler, to find the hot-spots

**Comment by davbeek** [comment]
We use BeautifulSoup (bs4), which uses internal data structures to process html, without actually using html. Only at the very end, html is generated by means of `html = str(soup)`, where `soup` is the bs4 object that represents the html.

**Comment by davbeek** [comment]
With the version .0.3.5 from Dec 2024, execution time for the main repo for the website went from 43.3s without multi-threading to 23.8 with multi-threading, so a reduction from 100% to 55%, even with a threading lock for repository access for the git log commands via GitPython.

**Comment by davbeek** [comment]
Closing.
